block/deserialize.go

```diff
@@ -0,0 +1,341 @@
+package block
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/vulpemventures/go-elements/internal/bufferutil"
+
+	"github.com/vulpemventures/go-elements/transaction"
+)
+
+func deserialize(buf *bytes.Buffer) (*Block, error) {
+	header, err := DeserializeHeader(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	transactions, err := DeserializeTransactions(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Block{
+		Header: header,
+		TransactionsData: &Transactions{
+			Transactions: transactions,
+		},
+	}, nil
+}
+
+func DeserializeHeader(
+	buf *bytes.Buffer,
+) (*Header, error) {
+	d := bufferutil.NewDeserializer(buf)
+
+	version, err := d.ReadUint32()
+	if err != nil {
+		return nil, err
+	}
+
+	isDyna := false
+	if version>>31 == 1 {
+		isDyna = true
+	}
+	version = blockVersion
+
+	prevBlockHash, err := d.ReadSlice(hashSize)
+	if err != nil {
+		return nil, err
+	}
+
+	merkleRoot, err := d.ReadSlice(hashSize)
+	if err != nil {
+		return nil, err
+	}
+
+	timestamp, err := d.ReadUint32()
+	if err != nil {
+		return nil, err
+	}
+
+	blockHeight, err := d.ReadUint32()
+	if err != nil {
+		return nil, err
+	}
+
+	extData, err := deserializeExtData(d, isDyna)
+	if err != nil {
+		return nil, err
+	}
+
+	extData.IsDyna = isDyna
+
+	return &Header{
+		Version:       version,
+		PrevBlockHash: prevBlockHash,
+		MerkleRoot:    merkleRoot,
+		Timestamp:     timestamp,
+		Height:        blockHeight,
+		ExtData:       extData,
+	}, nil
+}
+
+func deserializeExtData(
+	d *bufferutil.Deserializer,
+	isDyna bool,
+) (*ExtData, error) {
+	var dynamicFederation *DynamicFederation
+	var proof *Proof
+	var err error
+	if isDyna {
+		dynamicFederation, err = deserializeDynamicFederation(d)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		proof, err = deserializeProof(d)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &ExtData{
+		Proof:             proof,
+		DynamicFederation: dynamicFederation,
+	}, nil
+}
+
+func deserializeDynamicFederation(
+	d *bufferutil.Deserializer,
+) (*DynamicFederation, error) {
+	currentParams, err := deserializeDynamicFederationParams(d)
+	if err != nil {
+		return nil, err
+	}
+
+	proposedParams, err := deserializeDynamicFederationParams(d)
+	if err != nil {
+		return nil, err
+	}
+
+	signBlockWitness, err := deserializeSignBlockWitness(d)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DynamicFederation{
+		Current:          currentParams,
+		Proposed:         proposedParams,
+		SignBlockWitness: signBlockWitness,
+	}, nil
+}
+
+func deserializeDynamicFederationParams(
+	d *bufferutil.Deserializer,
+) (*DynamicFederationParams, error) {
+	var dynamicFederationParams *DynamicFederationParams
+	var compactParams *CompactParams
+	var fullParams *FullParams
+	var err error
+	nul := false
+
+	serializeType, err := d.ReadUint8()
+	if err != nil {
+		return nil, err
+	}
+
+	switch serializeType {
+	case null:
+		nul = true
```

Comment:
why not directly `return nil, nil` here so we can get rid of this `nul` variable? If I didn't get it wrong, in case serializeType is null, dynamicFederationParams is nil and it's returned just like that, isn't it?

Reply:
Done